extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java

```diff
@@ -227,7 +257,15 @@ private String countQuery(String selectQuery) {
     public <T extends Entity> List<T> list() {
         Query jpaQuery = createQuery();
         try (NonThrowingCloseable c = applyFilters()) {
-            return jpaQuery.getResultList();
+            List<T> results = jpaQuery.getResultList();
+            if (resultMapper != null) {
+                List<T> mappedResults = new ArrayList<>(results.size());
+                for (Object result : results) {
+                    mappedResults.add((T) resultMapper.apply(result));
+                }
+                results = mappedResults;
+            }
+            return results;
```

Comment:
Something seems wrong here are you map a type `T` to another type `T`.
It works due to the fact that generics are erased at runtime but it looks like someting is not correct. 
Maybe it's linked to the other remarks I noticed about `Object` being used in the signature of the resultMapper.

`List<T> results = jpaQuery.getResultList();` is not correct as `getResultList();` returns a raw List, so either use `List<Object>` or `List<?>`.

Reply:
So, I think a more correct code and understandable code should look like:

```
if (resultMapper == null) return jpaQuery.getResultList();
List results = jpaQuery.getResultList();
List<T> mappedResults = new ArrayList<>(results.size());
for (Object result : results) {
    mappedResults.add((T) resultMapper.apply(result));
}
return mappedResults;
```

Is the cast to `T` needed ? The resutMapper should have the correct return type if generics are correctly handled right ?

Reply:
Yes, the `List<T>` should be just `List` because we map `Object` to `T`. Anyway, I'm just in the middle of renaming and refactoring. So let's discuss the final code...

Reply:
BTW the `List<T> results = jpaQuery.getResultList()` is used all around this class...

Reply:
Generics hell...

Reply:
> Is the cast to T needed ?

It is for some reason. At least in my Eclipse IDE.